File: Taboola SDK ObjC Samples/Taboola SDK ObjC Samples/SDK API/Feed/FeedViewController.m
//
//  FeedViewController.m
//  Taboola API ObjC Samples
//
//  Created by Roman Slyepko on 1/17/19.
//  Copyright © 2019 Taboola. All rights reserved.
//

#import "FeedViewController.h"
#import "FeedTableViewCell.h"
#import <TaboolaSDK/TaboolaSDK.h>

@interface FeedViewController () <TaboolaApiClickDelegate>
@property (nonatomic) TaboolaApi *taboolaApi;
@property (nonatomic) TBRecommendationRequest *recomendationRequest;
@property (nonatomic) NSMutableArray *itemsArray;
@property (nonatomic) TBPlacement *placement;
@end

@implementation FeedViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.itemsArray = [NSMutableArray new];
    
    self.taboolaApi = [TaboolaApi sharedInstance];
    [self.taboolaApi startWithPublisherID:@"sdk-tester-demo" andApiKey:@"30dfcf6b094361ccc367bbbef5973bdaa24dbcd6"];
    self.taboolaApi.clickDelegate = self;
    
    [self fetchRecommendation];
}

- (void)fetchRecommendation {
    self.recomendationRequest = [TBRecommendationRequest new];
    self.recomendationRequest.sourceType = TBSourceTypeText;
    [self.recomendationRequest setPageUrl: @"http://www.example.com"];
    
    TBPlacementRequest *parameters = [TBPlacementRequest new];
    parameters.name = @"article";
    parameters.recCount = 2;
    
    [self.recomendationRequest addPlacementRequest:parameters];
    
    __weak FeedViewController *weakSelf = self;
    [self.taboolaApi fetchRecommendations:self.recomendationRequest onSuccess:^(TBRecommendationResponse *response) {
        
        TBPlacement *placement = response.placements.firstObject;
        
        weakSelf.placement = placement;
        weakSelf.itemsArray = [placement.listOfItems mutableCopy];
        [weakSelf.tableView reloadData];
        
    } onFailure:^(NSError *error) {
        NSLog(@"Taboola API: error fetching recommendations: \n%@", error.localizedDescription);
    }];
}

- (void)fetchNextPage {
    if (self.placement == nil) {
        return;
    }
    __weak FeedViewController *weakSelf = self;
    [self.taboolaApi getNextBatchForPlacement:self.placement itemsCount:3 onSuccess:^(TBRecommendationResponse *response) {
        TBPlacement *placement = response.placements.firstObject;
        weakSelf.placement = placement;
        [weakSelf.itemsArray addObjectsFromArray:placement.listOfItems];
        [weakSelf.tableView reloadData];
    } onFailure:^(NSError *error) {
        
    }];
}

-(void)dealloc {
    [self.taboolaApi clear];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.itemsArray count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    NSString* identifier = @"FeedCell";
    FeedTableViewCell* cell = [tableView dequeueReusableCellWithIdentifier:identifier];
    
    TBItem *item = self.itemsArray[indexPath.row];
    [item initThumbnailView:cell.mainImageView];
    [item initTitleView:cell.titleView];
    [item initBrandingView:cell.brandingView];
    return cell;
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row == self.itemsArray.count - 1) {
        [self fetchNextPage];
    }
}

#pragma mark - Taboola API Click Delegate

- (BOOL)onItemClick:(NSString *)placementName withItemId:(NSString *)itemId withClickUrl:(NSString *)clickUrl isOrganic:(BOOL)organic {
    return false;
}

@end

File: Taboola SDK Swift Sample/Taboola SDK Swift Sample/SDK API/Feed/FeedViewController.swift
//
//  FeedViewController.swift
//  Taboola API Swift Sample
//
//  Created by Roman Slyepko on 2/15/19.
//  Copyright © 2019 Taboola. All rights reserved.
//

import UIKit
import TaboolaSDK

class FeedViewController: UITableViewController {
    var recommendationRequest: TBRecommendationRequest?
    var placement: TBPlacement?
    
    var items = [TBItem]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        TaboolaApi.sharedInstance()?.start(withPublisherID: "sdk-tester", andApiKey: "d39df1418f5a4819c9eae2ca02595d57de98c246")
        TaboolaApi.sharedInstance()?.clickDelegate = self
        
        fetchRecommendations()
    }

    func fetchRecommendations() {
        recommendationRequest = TBRecommendationRequest()
        recommendationRequest?.sourceType = TBSourceTypeText
        recommendationRequest?.setPageUrl("http://www.example.com")
        
        let parameters = TBPlacementRequest()
        parameters.name = "article"
        parameters.recCount = 2
        
        recommendationRequest?.add(parameters)
        
        TaboolaApi.sharedInstance()?.fetchRecommendations(recommendationRequest,
                                                          onSuccess: {[weak self] (response) in
                                                            guard let placement = response?.placements.firstObject as? TBPlacement else { return }
                                                            self?.placement = placement
                                                            self?.items = placement.listOfItems
                                                            self?.tableView.reloadData()
        }, onFailure: { (error) in
            print("Taboola API: error fetching recommendations: \(error?.localizedDescription)")
        })
    }
    
    func fetchNextPage() {
        guard let placement = placement else {
            return
        }
        TaboolaApi.sharedInstance()?.getNextBatch(for: placement, itemsCount: 3, onSuccess: {[weak self] (response) in
            self?.placement = placement
            self?.items.append(contentsOf: placement.listOfItems)
            self?.tableView.reloadData()
        }, onFailure: { (error) in
            print("Taboola API: error fetching next page: \(error?.localizedDescription)")
        })
    }
    
    //MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return items.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "FeedCell") as? FeedTableViewCell ?? FeedTableViewCell()
        let item = items[indexPath.row]
        item.initTitleView(cell.titleView)
        item.initThumbnailView(cell.mainImageView)
        item.initBrandingView(cell.brandingView)
        item.initDescriptionView(cell.descriptionView)
        return cell
    }
    
    override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        if indexPath.row == items.count - 1 {
            fetchNextPage()
        }
    }

}

extension FeedViewController: TaboolaApiClickDelegate {
    func onItemClick(_ placementName: String!, withItemId itemId: String!, withClickUrl clickUrl: String!, isOrganic organic: Bool) -> Bool {
        return true
    }
    
    
}
